src/bots/liquidator.ts

```diff
@@ -217,17 +219,147 @@ export class PerpLiquidatorBot implements Bot {
+	private findPerpBankruptingMarket(
+		chUserToCheck: ClearingHouseUser
+	): [number, BN] {
+		let largestPositionLossMarketIndex = -1;
+		let largestPositionLoss = ZERO;
+		for (const market of this.clearingHouse.getPerpMarketAccounts()) {
+			const position = chUserToCheck.getUserPosition(market.marketIndex);
+			if (position.quoteAssetAmount.lt(largestPositionLoss)) {
+				largestPositionLossMarketIndex = market.marketIndex;
+				largestPositionLoss = position.quoteAssetAmount;
+			}
+		}
+
+		return [largestPositionLossMarketIndex, largestPositionLoss];
+	}
+
+	/**
+	 * Find the user spot position with the largest loss, resolve bankruptcy on this market.
+	 *
+	 * @param chUserToCheck
+	 * @returns
+	 */
+	private findSpotBankruptingMarket(
+		chUserToCheck: ClearingHouseUser
+	): [number, BN] {
+		let largestPositionLossMarketIndex = -1;
+		let largestPositionLoss = ZERO;
+
+		// TODO: use getSpotMarketAccounts once it's merged to sdk
+		// for (const market of this.clearingHouse.getSpotMarketAccounts()) {
+		// 	const position = chUserToCheck.getSpotPosition(market.marketIndex);
+		// 	if (position.quoteAssetAmount.lt(largestPositionLoss)) {
+		// 		largestPositionLossMarketIndex = market.marketIndex;
+		// 		largestPositionLoss = position.quoteAssetAmount;
+		// 	}
+		// }
+
+		// temp on old sdk impl.
+		for (const position of chUserToCheck.getUserAccount().spotPositions) {
+			if (!isVariant(position.balanceType, 'borrow')) {
+				// not possible to resolve non-borrow markets
+				continue;
+			}
+			const spotMarket = this.clearingHouse.getSpotMarketAccount(
+				position.marketIndex
+			);
+			const positionBorrowAmount = getTokenAmount(
+				position.scaledBalance,
+				spotMarket,
+				position.balanceType
+			);
+
+			// spot borrow is uint
+			if (positionBorrowAmount.gt(largestPositionLoss)) {
+				largestPositionLoss = positionBorrowAmount;
+				largestPositionLossMarketIndex = position.marketIndex;
+			}
+		}
+
+		return [largestPositionLossMarketIndex, largestPositionLoss];
+	}
+
+	private async tryResolveBankruptUser(user: ClearingHouseUser) {
+		const userAcc = user.getUserAccount();
+		const userKey = user.getUserAccountPublicKey();
+		const auth = userAcc.authority.toBase58();
+
+		// find out whether the user is perp-bankrupt or spot-bankrupt
+		const [perpIdx, perpLoss] = this.findPerpBankruptingMarket(user);
+		const [spotIdx, spotLoss] = this.findSpotBankruptingMarket(user);
+
+		// resolve the largest loss market
+		let marketTypeToResolve = '';
+		if (perpIdx === -1 && spotIdx === -1) {
+			// no market to resolve
+			logger.warn(
+				`UserAcc: ${userKey.toBase58()}, Auth: ${auth} is bankrupt but no resolvable market`
+			);
+		} else if (perpIdx > -1 && spotIdx === -1) {
+			marketTypeToResolve = 'perp';
+		} else if (perpIdx === -1 && spotIdx > -1) {
+			marketTypeToResolve = 'spot';
+		} else {
+			// TODO: are perpLoss and spotLoss precisions correct to compare here?
+			if (perpLoss.abs().gt(spotLoss.abs())) {
```

Comment:
ok to do if you cast it into usd, but you may just want to try to resolve them both at the same time? why'd you want to do it one at a time based on size?

Reply:
probably got a bit mixed up with how liquidation worked, so thought i should resolve 1 position, then check if the account is still bankrupting and resolve the next. going to just resolve all markets with loss as Z pointed out too